src/state/CallViewModel.ts

```diff
@@ -576,34 +641,61 @@ export class CallViewModel extends ViewModel {
     }),
   );
 
-  private readonly spotlightAndPip: Observable<
-    [Observable<MediaViewModel[]>, Observable<UserMediaViewModel | null>]
-  > = this.screenShares.pipe(
-    map((screenShares) =>
-      screenShares.length > 0
-        ? ([of(screenShares.map((m) => m.vm)), this.spotlightSpeaker] as const)
-        : ([
-            this.spotlightSpeaker.pipe(map((speaker) => [speaker!])),
-            this.spotlightSpeaker.pipe(
-              switchMap((speaker) =>
-                speaker.local
-                  ? of(null)
-                  : this.localUserMedia.pipe(
-                      switchMap((vm) =>
-                        vm.alwaysShow.pipe(
-                          map((alwaysShow) => (alwaysShow ? vm : null)),
-                        ),
-                      ),
-                    ),
-              ),
-            ),
-          ] as const),
-    ),
-  );
-
   private readonly spotlight: Observable<MediaViewModel[]> =
-    this.spotlightAndPip.pipe(
-      switchMap(([spotlight]) => spotlight),
+    this.screenShares.pipe(
+      switchMap((screenShares) => {
+        if (screenShares.length > 0) {
+          return of(screenShares.map((m) => m.vm));
+        }
+
+        return this.spotlightSpeaker.pipe(
+          map((speaker) => (speaker ? [speaker] : [])),
+        );
+      }),
+      this.scope.state(),
+    );
+
+  private readonly pip: Observable<UserMediaViewModel | null> =
+    this.screenShares.pipe(
+      switchMap((screenShares) => {
+        if (screenShares.length > 0) {
+          return this.spotlightSpeaker;
+        }
+
+        return this.spotlightSpeaker.pipe(
+          switchMap((speaker) => {
+            if (!speaker || speaker.local) {
+              return of(null);
+            }
+
+            return this.mediaItems.pipe(
+              switchMap((mediaItems) => {
+                const localUserMedia = mediaItems.find(
+                  (m) => m.vm instanceof LocalUserMediaViewModel,
+                ) as UserMedia | undefined;
+
+                const localUserMediaViewModel = localUserMedia?.vm as
+                  | LocalUserMediaViewModel
+                  | undefined;
+
+                if (!localUserMediaViewModel) {
+                  return of(null);
+                }
+
+                return localUserMediaViewModel.alwaysShow.pipe(
+                  map((alwaysShow) => {
+                    if (alwaysShow) {
+                      return localUserMediaViewModel;
+                    }
+
+                    return null;
+                  }),
+                );
+              }),
+            );
+          }),
+        );
+      }),
       this.scope.state(),
     );
```

Comment:
Trying to follow that code I noticed that I am not a big fan of nested switchMaps.
Its hard to follow when things update:
 - the main switch map is done on Screenshares so each screenShare update does the condition and also reinitializes the observable generated by `sportlightSpeaker.pipe`.
 -  which then can give us the newly reconstructed of(null) or another observable based on `this.mediaItems.pipe`.

 It does make sense when thinking it through. I still would like to propose this alternative where we first combine our dependent observables.
 I like to see the clear list:`this.screenShares, this.spotlightSpeaker, this.mediaItems` of dependent observables in one place.
 It of course now will execute the whole logic more often. `screenShares.length > 0` will also be checked if mediaItems change.
 But the logic is easier to follow since we do not have the overhead that sections of the if tree are not reevaluated.
```suggestion
    private readonly pip: Observable<UserMediaViewModel | null> = combineLatest([
    this.screenShares,
    this.spotlightSpeaker,
    this.mediaItems,
  ]).pipe(
    switchMap(([screenShares, spoltlight, mediaItems]) => {
      if (screenShares.length > 0) {
        return this.spotlightSpeaker;
      }
      if (!spoltlight || spoltlight.local) {
        return of(null);
      }

      const localUserMedia = mediaItems.find(
        (m) => m.vm instanceof LocalUserMediaViewModel,
      ) as UserMedia | undefined;

      const localUserMediaViewModel = localUserMedia?.vm as
        | LocalUserMediaViewModel
        | undefined;

      if (!localUserMediaViewModel) {
        return of(null);
      }
      return localUserMediaViewModel.alwaysShow.pipe(
        map((alwaysShow) => {
          if (alwaysShow) {
            return localUserMediaViewModel;
          }

          return null;
        }),
      );
    }),
    this.scope.state(),
  );
```

Reply:
That being said this is more then clear enough to merge it.
I am also curious about your take on this.

Reply:
As per offline conversation just now, I was tempted to refactor it into a `combineLatest` when I was working on it earlier in the week. If it does actually work as before then I think it is much more readable 👍